### PR TITLE
Moodle no apache

### PR DESCRIPTION
**Pull request by jvonau**
### Smoke-tested on which OS or OS's:
lightly test on ubuntu 20.10 to the point of logging in as Admin
Needs further shakedown but it's a start.

### CODE REVIEW

**Comment by holta** [comment]
Incredible news!

Hopefully this can converge quickly with diverse suggestions around what's most important to test?

Tangentially Related:

- #2829

**Comment by jvonau** [comment]
Testing routine:
- pull this PR into a new git branch
- `runrole moodle` should replace just the nginx config file, test the webpage should work but without css/js loading.
- `runrole --reinstall moodle` should add the missing group ownership to the moodle files, test the webpage to validate the ownership change.
-  poke around the site noting what may not work right.

**Comment by holta** [comment]
A custom `/etc/iiab/local_vars.yml` ([essentially Moodle only](https://paste.ubuntu.com/p/KCQys3NxXh/)) is now installing this PR onto Ubuntu Desktop 21.04 VM here:

197-u2104-desk-moodle-nginx-0701 = 10.8.0.62

Feel free to log in & reboot it as nec.

@deldesir definitely take a look if you too are interested.

**Comment by holta** [comment]
FYI this PR failed to install on 10.8.0.62 above.  Error pasted in below.

The fix should be straightforward &mdash; likely https://github.com/jvonau/iiab/blob/moodle-no-apache/roles/moodle/tasks/install.yml#L20-L31 should be commented out &mdash; so I'll give that a shot.

> TASK [moodle : Enact the equivalent of 'nginx_high_php_limits: True' in /etc/php                                                                                        /7.4/apache2/php.ini for Moodle 3.11+] ***
failed: [127.0.0.1] (item={'regexp': '^upload_max_filesize', 'line': 'upload_max                                                                                        _filesize = 500M    ; default is 2M'}) => {"ansible_loop_var": "item", "changed"                                                                                        : false, "item": {"line": "upload_max_filesize = 500M    ; default is 2M", "rege                                                                                        xp": "^upload_max_filesize"}, "msg": "Destination /etc/php/7.4/apache2/php.ini d                                                                                        oes not exist !", "rc": 257}
failed: [127.0.0.1] (item={'regexp': '^post_max_size', 'line': 'post_max_size =                                                                                         500M    ; default is 8M'}) => {"ansible_loop_var": "item", "changed": false, "it                                                                                        em": {"line": "post_max_size = 500M    ; default is 8M", "regexp": "^post_max_si                                                                                        ze"}, "msg": "Destination /etc/php/7.4/apache2/php.ini does not exist !", "rc":                                                                                         257}
failed: [127.0.0.1] (item={'regexp': '^max_execution_time', 'line': 'max_executi                                                                                        on_time = 300    ; default is 30'}) => {"ansible_loop_var": "item", "changed": f                                                                                        alse, "item": {"line": "max_execution_time = 300    ; default is 30", "regexp":                                                                                         "^max_execution_time"}, "msg": "Destination /etc/php/7.4/apache2/php.ini does no                                                                                        t exist !", "rc": 257}
failed: [127.0.0.1] (item={'regexp': '^max_input_time', 'line': 'max_input_time                                                                                         = 300    ; default is 60'}) => {"ansible_loop_var": "item", "changed": false, "i                                                                                        tem": {"line": "max_input_time = 300    ; default is 60", "regexp": "^max_input_                                                                                        time"}, "msg": "Destination /etc/php/7.4/apache2/php.ini does not exist !", "rc"                                                                                        : 257}
failed: [127.0.0.1] (item={'regexp': '^memory_limit', 'line': 'memory_limit = 51                                                                                        2M    ; default is 128M / Nextcloud requests 512M'}) => {"ansible_loop_var": "it                                                                                        em", "changed": false, "item": {"line": "memory_limit = 512M    ; default is 128                                                                                        M / Nextcloud requests 512M", "regexp": "^memory_limit"}, "msg": "Destination /e                                                                                        tc/php/7.4/apache2/php.ini does not exist !", "rc": 257}
failed: [127.0.0.1] (item={'regexp': '^max_input_vars', 'line': 'max_input_vars                                                                                         = 5000    ; default is 1000 / Moodle 3.11+ requires 5000+ with PHP 8+'}) => {"an                                                                                        sible_loop_var": "item", "changed": false, "item": {"line": "max_input_vars = 50                                                                                        00    ; default is 1000 / Moodle 3.11+ requires 5000+ with PHP 8+", "regexp": "^                                                                                        max_input_vars"}, "msg": "Destination /etc/php/7.4/apache2/php.ini does not exis                                                                                        t !", "rc": 257}

**Comment by holta** [comment]
As amended, this PR installed onto 10.8.0.62 (Ubuntu Desktop 21.04).

However visiting http://box.lan/moodle on this very same machine (VM) shows this error in the browser:

> 404 Not Found
> nginx/1.18.0 (Ubuntu)

I also tried http://box.lan/moodle/dataroot which shows the same error.

iiab-diagnostics output: http://sprunge.us/6IKNgx?en

Please log in to 10.8.0.62 directly if that helps.

**Comment by holta** [comment]
FYI I just rebooted 10.8.0.62 for the heck of it, to see if anything might change.

Result: http://box.lan/moodle still fails (404 error).

**Comment by jvonau** [comment]
Think you forgot about moodle's locked url, edit /opt/iiab/moodle/config.php's wwwroot= to match the ip address you are trying to access. I don't think your VM has wifi to test box.lan. That is why I suggested a configured machine to try this with.

**Comment by holta** [comment]
> Think you forgot about moodle's locked url, edit /opt/iiab/moodle/config.php's wwwroot= to match the ip address you are trying to access. I don't think your VM has wifi to test box.lan. That is why I suggested a configured machine to try this with.

For sure the VM has no WiFi.

But this works:

```
root@box:~# ping box.lan
PING box.lan (127.0.0.1) 56(84) bytes of data.
64 bytes from box.lan (127.0.0.1): icmp_seq=1 ttl=64 time=0.046 ms
```

And `/opt/iiab/moodle/config.php` seems to be correct: (or at least normal!)

```
root@box:~# grep wwwroot /opt/iiab/moodle/config.php
$CFG->wwwroot   = 'http://box.lan/moodle';
```

Anything else to try?

**Comment by jvonau** [comment]
> > Think you forgot about moodle's locked url, edit /opt/iiab/moodle/config.php's wwwroot= to match the ip address you are trying to access. I don't think your VM has wifi to test box.lan. That is why I suggested a configured machine to try this with.
> 
> For sure the VM has no WiFi.
> 
> But this works:
> 

.... Puts on teaching hat

> ```
> root@box:~# ping box.lan
> PING box.lan (127.0.0.1) 56(84) bytes of data.
> 64 bytes from box.lan (127.0.0.1): icmp_seq=1 ttl=64 time=0.046 ms
The client machine you are trying from has host file entries for box.lan pointing to localhost. Which machine did you run this command on?


> And `/opt/iiab/moodle/config.php` seems to be correct: (or at least normal!)

It is normal, if you had wifi and accessing moodle over br0 aka box.lan.
> root@box:~# grep wwwroot /opt/iiab/moodle/config.php
> $CFG->wwwroot   = 'http://box.lan/moodle';
> ```
> 
> Anything else to try?

Edit wwwroot http://box.lan/moodle to be http://192.168.0.197/moodle in /opt/iiab/moodle/config.php

**Comment by jvonau** [comment]
https://github.com/iiab/iiab/issues/941#issuecomment-410792263

**Comment by holta** [comment]
Editing `/opt/iiab/moodle/config.php` does not help, regardless if this line...

```
$CFG->wwwroot   = 'http://box.lan/moodle';
```

...is changed to...

```
$CFG->wwwroot   = 'http://192.168.0.197/moodle';
```

...or is changed to...

```
$CFG->wwwroot   = 'http://10.8.0.62/moodle';
```

The result is always the same:

```
404 Not Found
nginx/1.18.0 (Ubuntu)
```

(All tests were done using a browser within the VM itself.  The VM was rebooted right after editing `config.php` every time, prior to testing.)

**Comment by jvonau** [comment]
> (All tests were done using a browser within the VM itself.

then use localhost or 127.0.0.1 for wwwroot

**Comment by holta** [comment]
These 2 lines (one at a time) were also attempted in `/opt/iiab/moodle/config.php` on 10.8.0.62 = 197-u2104-desk-moodle-nginx-0701:

```
$CFG->wwwroot   = 'http://localhost/moodle';
```

```
$CFG->wwwroot   = 'http://127.0.0.1/moodle';
```

The result (when trying to browse to those URL's) is the same:

```
404 Not Found
nginx/1.18.0 (Ubuntu)
```

(All tests were done using a browser within the VM itself. The VM was rebooted right after editing `config.php` every time, prior to testing.)

**Comment by jvonau** [comment]
The wwwroot will be the only address that should work from the client browser, ie if you set wwwroot to http://192.168.0.197/moodle that should work from a network client that can access 192.168.0.197 or from the local VM machine at the same url, all others addresses are known not to work on anything other than what wwwroot is set to. Pick on target interface and only test from clients that can access the ip address associated with that interface.

**Comment by jvonau** [comment]
post install look of moodle-ngnix.conf might of messed up the softcoding and one could try the attached file after renaming first take directly from my running install
```
location ~ ^/moodle/dataroot/ {
    internal;
    alias /library/moodle;
}

location ~ ^/moodle(.*)\.php(.*)$ {
    alias /opt/iiab/moodle$1.php$2;
    include fastcgi.conf;
    fastcgi_split_path_info ^(.+\.php)(/.+)$;
    fastcgi_read_timeout 300;
    fastcgi_pass php;
    fastcgi_index index.php;
    fastcgi_param   PATH_INFO        $fastcgi_path_info;
    fastcgi_param   SCRIPT_FILENAME  $document_root/$fastcgi_script_name;
    fastcgi_param   PHP_VALUE "max_execution_time=300\n upload_max_filesize=500M\n post_max_size=500M\n max_input_vars=5000";
}


location ~ ^/moodle/ {
    root /opt/iiab/;
}
```
[moodle-nginx.conf.txt](https://github.com/iiab/iiab/files/6750349/moodle-nginx.conf.txt)
$CFG->wwwroot   = 'http://localhost:2080/moodle';  I'm running nginx on an alternate port testing #2834 after logging in:
moodle-nginx

**Comment by jvonau** [comment]
set wwwroot to http://10.8.0.62/moodle and accessed the same url:

>2021/07/01 16:03:18 [info] 2717#2717: Using 131072KiB of shared memory for nchan in /etc/nginx/nginx.conf:104
2021/07/01 16:05:21 [error] 2721#2721: *1 open() "/library/www/html/moodle" failed (2: No such file or directory), client: 10.8.0.22, server: box, request: "GET /moodle HTTP/1.1", host: "10.8.0.62"

This is a list of active pages on my instance where it is working. ls /etc/iiab-nginx/conf.d/
> iiab.conf                   kolibri-nginx.conf  munin24-nginx.conf
internetarchive-nginx.conf  lokole-nginx.conf   nodered-nginx.conf
kiwix-nginx.conf            moodle-nginx.conf   osm-vector-maps-nginx.conf

What is perplexing is the 'root /opt/iiab' should override 'root /library/www/html/' in iiab.conf but is not and should as osm-vector-maps and wordpress fiddle with the root assignment also. This nginx config was written with osm-vector-maps in place perhaps that is what the difference is..

**Comment by jvonau** [comment]
Install osm-vector-maps, now works across the vpn when wwwroot is set to http://10.8.0.62/moodle

Screenshot at 2021-07-01 15-41-03

But the question is why...

**Comment by holta** [comment]
> Install osm-vector-maps, now works across the vpn when wwwroot is set to http://10.8.0.62/moodle

http://10.8.0.62/moodle is not working for me (404 error).  I've tried from 2 different machines, both on OpenVPN, and tested different browsers.  Is it working reliably for you?

**Comment by jvonau** [comment]
I can get past logging in..... caching/cookies?
Screenshot at 2021-07-01 16-08-41

**Comment by holta** [comment]
http://10.8.0.62/moodle 404's for me every time, from different browsers and machines on OpenVPN.

**Comment by jvonau** [comment]
get others to try across the vpn... works here every time.. has to be client caching...

**Comment by jvonau** [comment]
```
10.8.0.22 - - [01/Jul/2021:17:14:50 -0400] "GET /moodle/theme/font.php/boost/core/1625158508/fontawesome-webfont.woff2?v=4.7.0 HTTP/1.1" 200 77160 "http://10.8.0.62/moodle/theme/styles.php/boost/1625158508_1625158564/all" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:50 -0400] "GET /moodle/theme/image.php/boost/theme/1625158508/favicon HTTP/1.1" 200 1150 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:51 -0400] "GET /moodle/lib/requirejs.php/1625158508/core/first.js HTTP/1.1" 200 308038 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:51 -0400] "GET /moodle/lib/javascript.php/1625158508/lib/jquery/jquery-3.5.1.min.js HTTP/1.1" 200 30958 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:52 -0400] "GET /moodle/theme/yui_combo.php?3.17.2/event-mousewheel/event-mousewheel-min.js&3.17.2/event-resize/event-resize-min.js&3.17.2/event-hover/event-hover-min.js&3.17.2/event-touch/event-touch-min.js&3.17.2/event-move/event-move-min.js&3.17.2/event-flick/event-flick-min.js&3.17.2/event-valuechange/event-valuechange-min.js&3.17.2/event-tap/event-tap-min.js HTTP/1.1" 200 4827 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:52 -0400] "POST /moodle/lib/ajax/service.php?sesskey=BlmES5dR4X&info=media_videojs_get_language HTTP/1.1" 200 4549 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:52 -0400] "GET /moodle/lib/ajax/service-nologin.php?info=core_output_load_fontawesome_icon_system_map&cachekey=1625158508&args=%5B%7B%22index%22%3A0%2C%22methodname%22%3A%22core_output_load_fontawesome_icon_system_map%22%2C%22args%22%3A%7B%22themename%22%3A%22boost%22%7D%7D%5D HTTP/1.1" 200 30379 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:52 -0400] "GET /moodle/lib/ajax/service-nologin.php?info=core_output_load_template_with_dependencies,core_output_load_template_with_dependencies,core_output_load_template_with_dependencies,core_output_load_template_with_dependencies&cachekey=1625158508&args=%5B%7B%22index%22%3A0%2C%22methodname%22%3A%22core_output_load_template_with_dependencies%22%2C%22args%22%3A%7B%22component%22%3A%22core%22%2C%22template%22%3A%22loading%22%2C%22themename%22%3A%22boost%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A1%2C%22methodname%22%3A%22core_output_load_template_with_dependencies%22%2C%22args%22%3A%7B%22component%22%3A%22core%22%2C%22template%22%3A%22modal%22%2C%22themename%22%3A%22boost%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A2%2C%22methodname%22%3A%22core_output_load_template_with_dependencies%22%2C%22args%22%3A%7B%22component%22%3A%22core%22%2C%22template%22%3A%22modal_backdrop%22%2C%22themename%22%3A%22boost%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A3%2C%22methodname%22%3A%22core_output_load_template_with_dependencies%22%2C%22args%22%3A%7B%22component%22%3A%22core%22%2C%22template%22%3A%22pix_icon_fontawesome%22%2C%22themename%22%3A%22boost%22%2C%22lang%22%3A%22en%22%7D%7D%5D HTTP/1.1" 200 2434 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.22 - - [01/Jul/2021:17:14:52 -0400] "GET /moodle/lib/ajax/service-nologin.php?info=6-method-calls&cachekey=1625158562&args=%5B%7B%22index%22%3A0%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22cancel%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A1%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22closebuttontitle%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A2%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22loading%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A3%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22savechanges%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A4%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22showless%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core_form%22%2C%22lang%22%3A%22en%22%7D%7D%2C%7B%22index%22%3A5%2C%22methodname%22%3A%22core_get_string%22%2C%22args%22%3A%7B%22stringid%22%3A%22showmore%22%2C%22stringparams%22%3A%5B%5D%2C%22component%22%3A%22core_form%22%2C%22lang%22%3A%22en%22%7D%7D%5D HTTP/1.1" 200 222 "http://10.8.0.62/moodle/" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
192.168.0.197 - - [01/Jul/2021:17:16:39 -0400] "GET /moodle HTTP/1.1" 404 134 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.62 - - [01/Jul/2021:17:16:53 -0400] "GET /moodle HTTP/1.1" 404 134 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
127.0.0.1 - - [01/Jul/2021:17:17:02 -0400] "GET /moodle HTTP/1.1" 404 134 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
127.0.0.1 - - [01/Jul/2021:17:17:07 -0400] "GET /sdws HTTP/1.1" 404 134 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
```

**Comment by holta** [comment]
> has to be client caching...

Browser caches were completely cleared, and http://10.8.0.62/moodle still 404's on each machine.

I have no idea why.

**Comment by holta** [comment]
I've confirmed this by installing a new browser on 10.8.0.62 itself.

This new browser (Chromium, never installed before on this VM) also 404's when visiting http://10.8.0.62/moodle

**Comment by jvonau** [comment]
add a trailing slash?
10.8.0.22 - - [01/Jul/2021:17:21:03 -0400] "GET /moodle/ HTTP/1.1" 200 7116 "http://10.8.0.62/moodle/login/change_password.php" "Mozilla/5.0 (X11; Ubuntu; Linux aarch64; rv:89.0) Gecko/20100101 Firefox/89.0" "http_x_forwarded_for"
10.8.0.62 - - [01/Jul/2021:17:36:45 -0400] "GET /moodle HTTP/1.1" 404 197 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36" "http_x_forwarded_for"

**Comment by holta** [comment]
> add a trailing slash?

That works!

(We'll have to bake that into moodle-nginx.conf eventually, to handle both cases.)

**Comment by jvonau** [comment]
yea the overall nginx infrastructure needs a review and improvement.

File: roles/moodle/templates/moodle-nginx.conf.j2 [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
+    fastcgi_read_timeout 300;
+    fastcgi_pass php;
+    fastcgi_index index.php;
+    fastcgi_param   PATH_INFO        $fastcgi_path_info;
+    fastcgi_param   SCRIPT_FILENAME   $document_root/$fastcgi_script_name;
+    fastcgi_param   PHP_VALUE "max_execution_time=300\n upload_max_filesize=500M\n post_max_size=500M\n max_input_vars=5000";
+}
+
+
+location ~ ^/moodle/ {
```
Comment (holta) [inline roles/moodle/templates/moodle-nginx.conf.j2]: This fix (appears) to solve the problem of http://box.lan/moodle 404's when the URL's trailing slash is missing.

```suggestion
location ~ ^/moodle {
```

Comment (jvonau) [reply roles/moodle/templates/moodle-nginx.conf.j2]: If restarting nginx doesn't puke, that seems reasonable

Comment (holta) [reply roles/moodle/templates/moodle-nginx.conf.j2]: This PR (but amended on Line 19 of moodle/templates/moodle-nginx.conf.j2) is being re-installed on a new VM 192.168.0.197 = 10.8.0.46 = 197-u2104-desk-moodle-nginx-0702

Using the same sub-minimal local_vars.yml ("Moodle only, nothing else") as yesterday.

To help evaluate the OSM theory, and others as necessary.

Comment (holta) [reply roles/moodle/templates/moodle-nginx.conf.j2]: Conclusion:

http://10.8.0.46/moodle and http://10.8.0.46/moodle/ both work without OSM (on above NEW "Moodle only" VM).

**Comment by holta** [comment]
Do the network settings look Ok hereunder?
http://10.8.0.46/moodle/admin/category.php?category=server

If so, this PR (amended for trailing slash) might be ready to merge?

FYI Moodle passes all its own checks, as confirmed by:
http://10.8.0.46/moodle/admin/environment.php

(Password sent to @jvonau privately, just write me if you too need the password!)

Tangentially related:

- PR #2840

**Pull request closed without merging**